File: practical 5 quick sort.cpp
#include <iostream>
#include <vector>
#include <ctime>
#include <cstdlib>

// Forward declaration of the partition function for deterministic quicksort
int deterministicPartition(std::vector<int> &numbers, int left, int right);

// Function to perform deterministic quicksort
void deterministicQuickSort(std::vector<int> &numbers, int left, int right) {
    if (left < right) {
        int partitionIndex = deterministicPartition(numbers, left, right);

        deterministicQuickSort(numbers, left, partitionIndex - 1);
        deterministicQuickSort(numbers, partitionIndex + 1, right);
    }
}

// Function to perform randomized quicksort
void randomizedQuickSort(std::vector<int> &numbers, int left, int right) {
    if (left < right) {
        int pivotIndex = left + rand() % (right - left + 1);

        std::swap(numbers[pivotIndex], numbers[right]);
        int pivot = numbers[right];
        int partitionIndex = deterministicPartition(numbers, left, right);

        randomizedQuickSort(numbers, left, partitionIndex - 1);
        randomizedQuickSort(numbers, partitionIndex + 1, right);
    }
}

// Function to partition the array for quicksort in deterministic quicksort
int deterministicPartition(std::vector<int> &numbers, int left, int right) {
    int pivot = numbers[left + (right - left) / 2];
    int i = left;
    int j = right;

    while (i <= j) {
        while (numbers[i] < pivot) {
            i++;
        }
        while (numbers[j] > pivot) {
            j--;
        }
        if (i <= j) {
            std::swap(numbers[i], numbers[j]);
            i++;
            j--;
        }
    }

    return i;
}

int main() {
    srand(time(0)); // Seed the random number generator

    std::vector<int> numbers(10000);
    for (int i = 0; i < numbers.size(); ++i) {
        numbers[i] = rand();
    }

    std::vector<int> deterministicSort = numbers; // Create a copy for deterministic sort
    std::vector<int> randomizedSort = numbers; // Create a copy for randomized sort

    clock_t deterministicTime = clock();
    deterministicQuickSort(deterministicSort, 0, deterministicSort.size() - 1);
    clock_t deterministicTimeEnd = clock();

    clock_t randomizedTime = clock();
    randomizedQuickSort(randomizedSort, 0, randomizedSort.size() - 1);
    clock_t randomizedTimeEnd = clock();

    std::cout << "Deterministic quick sort time: " << (double)(deterministicTimeEnd - deterministicTime) / CLOCKS_PER_SEC << " seconds\n";
    std::cout << "Randomized quick sort time: " << (double)(randomizedTimeEnd - randomizedTime) / CLOCKS_PER_SEC << " seconds\n";

    return 0;
}

File: practical 5 quick sort.java
import java.util.Random;

public class Main {

    public static void main(String[] args) {
        // Create an array of random numbers
        int[] numbers = new int[10000];
        Random rand = new Random();
        for (int i = 0; i < numbers.length; i++) {
            numbers[i] = rand.nextInt();
        }

        // Run the deterministic quick sort algorithm
        long deterministicTime = System.currentTimeMillis();
        deterministicQuickSort(numbers, 0, numbers.length - 1);
        long deterministicTimeEnd = System.currentTimeMillis();

        // Run the randomized quick sort algorithm
        long randomizedTime = System.currentTimeMillis();
        randomizedQuickSort(numbers, 0, numbers.length - 1, rand);
        long randomizedTimeEnd = System.currentTimeMillis();

        // Print the running times of the two algorithms
        System.out.println("Deterministic quick sort time: " + (deterministicTimeEnd - deterministicTime));
        System.out.println("Randomized quick sort time: " + (randomizedTimeEnd - randomizedTime));
    }

    public static void deterministicQuickSort(int[] numbers, int left, int right) {
        if (left >= right) {
            return;
        }

        int pivot = numbers[(left + right) / 2];
        int partition = partition(numbers, left, right, pivot);

        deterministicQuickSort(numbers, left, partition - 1);
        deterministicQuickSort(numbers, partition + 1, right);
    }

    public static void randomizedQuickSort(int[] numbers, int left, int right, Random rand) {
        if (left >= right) {
            return;
        }

        int pivotIndex = rand.nextInt(right - left + 1) + left;
        int pivot = numbers[pivotIndex];

        swap(numbers, pivotIndex, right);

        int partition = partition(numbers, left, right, pivot);

        randomizedQuickSort(numbers, left, partition - 1, rand);
        randomizedQuickSort(numbers, partition + 1, right, rand);
    }

    public static int partition(int[] numbers, int left, int right, int pivot) {
        int i = left;
        int j = right - 1;

        while (i <= j) {
            while (i <= j && numbers[j] > pivot) {
                j--;
            }

            while (i <= j && numbers[i] < pivot) {
                i++;
            }

            if (i <= j) {
                swap(numbers, i, j);
                i++;
                j--;
            }
        }

        return i;
    }

    public static void swap(int[] numbers, int i, int j) {
        int temp = numbers[i];
        numbers[i] = numbers[j];
        numbers[j] = temp;
    }
}
